sim/dex-data.ts

```diff
@@ -103,11 +103,11 @@ export class BasicEffect implements EffectData {
 	/** The full description for this effect. */
 	desc: string;
 	/**
-	 * Is this item/move/ability/pokemon nonstandard? True for effects
+	 * Is this item/move/ability/pokemon nonstandard? Specified for effects
 	 * that have no use in standard formats: made-up pokemon (CAP),
-	 * glitches (Missingno etc), and Pokestar pokemon.
+	 * glitches (Missingno etc), Pokestar pokemon, etc.
 	 */
-	isNonstandard: boolean;
+	isNonstandard: Nonstandard | false;
```

Comment:
```suggestion
	isNonstandard: Nonstandard | null;
```

(and elsewhere)

I've been trying to move away from the `| false` pattern.

Reply:
Becaus Typescript has had issues with it (thinking its `boolean`)?[ I don't really like `null` anywhere TBH](https://www.youtube.com/watch?v=PSGEjv3Tqo0&feature=youtu.be&t=9m21s), *maybe* its OK as a missing `Object` reference, and `false` seems to be more explicit and read better (`isNonstandard: false` is more clear then `null`).

I'm fine to switch this, I'm just curious about your reasoning?

Reply:
Because `false` here is a sentinel value, rather than an actual boolean (it's impossible for this to be `true`).

And nulls are the usual sentinel value for "there's nothing here". Between `null` and `undefined`, I went with `null` mostly because it was easier to type, especially since `T?` in JSDoc meant `T | null`, making it much easier to type than `T | undefined`.

`null` also allows `null` to be a non-default value – i.e. for `foo(null)` to be different from `foo()`. I guess this might be an anti-pattern though; I don't think we actually use this anywhere.

I don't think the difference between `null` and `undefined` is big enough to be worth redoing the entire codebase towards `undefined`.

Reply:
> Because `false` here is a sentinel value, rather than an actual boolean (it's impossible for this to be `true`). And nulls are the usual sentinel value for "there's nothing here"

Fair. This argument seems the most valid. My complaint is that https://github.com/Zarel/Pokemon-Showdown/pull/5330/commits/909eb214285b95ee9b2eaa760047bd95bffd9a1b#diff-a9f5618c5913fe6f011f25e44b509967 now doesn't read as well because there's an asymmetry between `isUnreleased` and `isNonstandard`, but I can look past that.

> I went with null mostly because it was easier to type, especially since `T?` in JSDoc meant `T | null`, making it much easier to type than `T | undefined`.

Yeah, unfortunately that is now the opposite in native Typescript. Instead of `T?` we now write `T | null = null` which is very unfortunate. 

> `null` also allows `null` to be a non-default value – i.e. for `foo(null)` to be different from `foo()`. I guess this might be an anti-pattern though; I don't think we actually use this anywhere.

I agree, this is definitely an anti pattern

> I don't think the difference between `null` and `undefined` is big enough to be worth redoing the entire codebase towards `undefined`.

I don't know, I'd consider it to clean up the `.ts` method signatures

Reply:
`null` is still significantly fewer letters than `undefined` overall, and there are many places you can't just use `?`, such as variable declarations. For the most part, I don't have a problem with `?` in method signatures; there are relatively few where the difference matters (and for those, we already have a defined difference between `null` and `undefined`).